### PR TITLE
pkg/obfuscate: add option to collect SQL comments and commands

### PR DESCRIPTION
**Pull request by alexbarksdale**
### What does this PR do?
Adds the option to collect SQL comments and commands during obfuscation which is configurable in `SQLConfig`. The agent python stub (CGo binding) is also updated to reflect these changes. Refer to **Additional Notes** for more information regarding the CGo bindings.

### Motivation
Comment collection: This change will enable us (Database Monitoring) to add trace ID propagation for SQL queries.
Command collection: This will enable us (Database Monitoring) to create facets for SQL commands so users can filter by them.

### Additional Notes

Some of our database integrations utilize the obfuscation library in the agent to obfuscate a SQL statement via [C bindings](https://github.com/DataDog/datadog-agent/blob/main/rtloader/common/builtins/datadog_agent.c#L681-L728). The C bindings invoke the [obfuscation library in Go](https://github.com/DataDog/datadog-agent/blob/alex.barksdale/collect-sql-metadata/pkg/collector/python/datadog_agent.go#L224-L260) when [this Python method](https://github.com/DataDog/integrations-core/blob/master/datadog_checks_base/datadog_checks/base/stubs/datadog_agent.py#L77-L79) is used.

Related PRs:
- Postgres integration update: https://github.com/DataDog/integrations-core/pull/10879
- MySQL integration update: https://github.com/DataDog/integrations-core/pull/10880
- SQLServer integration update: https://github.com/DataDog/integrations-core/pull/10881

### Describe how to test/QA your changes
- Run the tests in `pkg/obfuscate/sql_test.go`
- Run the agent stub tests by `$ invoke rtloader.test`
- Build the agent locally with the postgres integration and enable `dbm` in the integration instance config

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by alexbarksdale** [comment]
Do note that this PR **cannot** be merged until the following integrations are updated to support this.
- Postgres
- MySQL
- SQLServer

File: pkg/obfuscate/obfuscate.go [hunk 1]
```diff
@@ -124,6 +127,14 @@ type SQLConfig struct {
 	Cache bool
 }
 
+// SQLMetadata holds metadata collected throughout the obfuscation for a SQL statement.
```
Comment (gbbr) [inline pkg/obfuscate/obfuscate.go:130]:
```suggestion
// SQLMetadata holds metadata collected throughout the obfuscation of an SQL statement. It is only 
// collected when enabled via SQLConfig.
```

File: pkg/obfuscate/obfuscate.go [hunk 2]
```diff
@@ -105,7 +105,10 @@ type StatsClient interface {
 type SQLConfig struct {
 	// TableNames specifies whether the obfuscator should also extract the table names that a query addresses,
 	// in addition to obfuscating.
-	TableNames bool
+	TableNames bool `json:"table_names"`
+
+	// CollectComments specifies whether the obfuscator should also extract comments from the SQL statement.
```
Comment (gbbr) [inline pkg/obfuscate/obfuscate.go:110]:
```suggestion
	// CollectComments specifies whether the obfuscator should extract and return comments as SQL metadata when obfuscating.
```
Let's "hint" at the SQLMetadata struct.

File: pkg/obfuscate/sql.go [hunk 3]
```diff
@@ -299,10 +306,15 @@ func attemptObfuscation(tokenizer *SQLTokenizer) (*ObfuscatedQuery, error) {
...
 		lastToken       TokenKind
-		discard         = discardFilter{tokenizer.cfg.KeepSQLAlias}
-		replace         = replaceFilter{replaceDigits: tokenizer.cfg.ReplaceDigits}
-		grouping        groupingFilter
-		tableFinder     = tableFinderFilter{storeTableNames: storeTableNames}
+		metadata        SQLMetadata
+		discard         = discardFilter{
+			keepSQLAlias:    tokenizer.cfg.KeepSQLAlias,
+			collectComments: tokenizer.cfg.CollectComments,
+			metadata:        &metadata,
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:313]: Can we create a `metadataFinder` filter which extracts both the tables and the comments? Would that work? I think it makes more sense. This approach was quicker but a bit dirty IMHO.

Then, we could call something like `metadataFinder.Results()` which would return the `SQLMetadata`

**Review by gbbr — COMMENTED** [review]
Made a suggestion for improvement. I know it's more work, but I'm pretty sure it's much better, more maintainable and more readable.

File: pkg/obfuscate/sql.go [hunk 4]
```diff
@@ -18,7 +18,11 @@ var questionMark = []byte("?")
 
 // discardFilter is a token filter which discards certain elements from a query, such as
 // comments and AS aliases by returning a nil buffer.
-type discardFilter struct{ keepSQLAlias bool }
+type discardFilter struct {
+	keepSQLAlias    bool
+	collectComments bool
```
Comment (djmitche) [inline pkg/obfuscate/sql.go:23]: This seems to use `keep` and `collect` as equivalent -- would it make sense to stick to one term?

**Review by djmitche — APPROVED** [review]
`pkg/collector/python/datadog_agent.go` changes look good (for agent-core)

File: pkg/obfuscate/sql.go (hunk 4 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go:23]: Thanks for the review! I think `collect` makes more sense in this case because comments will be removed from the query and `keep` indicates that the comments will remain in the query after obfuscation.

Comment (djmitche) [reply pkg/obfuscate/sql.go:23]: Ah, nice distinction :)

File: pkg/obfuscate/sql.go (hunk 3 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go]: Thank you and great suggestion, this is indeed easier to read and maintain: https://github.com/DataDog/datadog-agent/pull/10042/commits/0522051bf95c902fb2a48a6099f4649d24985f27

File: pkg/obfuscate/sql.go [hunk 5]
```diff
@@ -16,9 +16,82 @@ import (
...
+type metadataFinderFilter struct {
+	collectTableNames bool
+	collectComments   bool
+
+	// comments keeps track of comments encountered by the filter.
+	comments []string
+	// tablesSeen keeps track of unique table names encountered by the filter.
+	tablesSeen map[string]struct{}
+	// tablesCsv specifies a comma-separated list of tables.
+	tablesCsv strings.Builder
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:30]:
```suggestion
	tablesCSV strings.Builder
```

**Review by gbbr — COMMENTED** [review]
It looks good now! Thanks! Please re-request a review from me once the dependencies of this change are met and you're ready to move forward.

File: pkg/obfuscate/sql.go (hunk 5 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go]: https://github.com/DataDog/datadog-agent/pull/10042/commits/14e376e06b86a665d0d8698e2009e3d88c44e428

File: pkg/obfuscate/sql.go [hunk 6]
```diff
@@ -16,9 +16,92 @@ import (
...
+}
+
+func (f *metadataFinderFilter) Filter(token, lastToken TokenKind, buffer []byte) (TokenKind, []byte, error) {
+	if f.collectComments && token == Comment {
+		// A comment with line-breaks will be brought to a single line.
+		f.comments = append(f.comments, strings.TrimSpace(strings.Replace(string(buffer), "\n", " ", -1)))
+	}
+	if f.collectCommands {
+		switch token {
+		case Select, Update, Insert, Join:
```
Comment (justiniso) [inline pkg/obfuscate/sql.go:43]: Delete, Alter, Drop, Create, Grant, Revoke, Commit, Begin, Truncate, With(if we want to count it, but I think not).

File: pkg/obfuscate/sql.go [hunk 7]
```diff
@@ -124,7 +207,7 @@ func (f *groupingFilter) Filter(token, lastToken TokenKind, buffer []byte) (toke
 	}
 
 	switch {
-	case f.groupMulti > 0 && lastToken == FilteredGroupableParenthesis && token == ID:
+	case f.groupMulti > 0 && lastToken == FilteredGroupableParenthesis && token == Select:
```
Comment (justiniso) [inline pkg/obfuscate/sql.go:210]: Huh, I don't think I follow. Can you explain why only Select is valid here?

File: pkg/obfuscate/obfuscate.go [hunk 8]
```diff
@@ -124,6 +130,17 @@ type SQLConfig struct {
 	Cache bool
 }
 
+// SQLMetadata holds metadata collected throughout the obfuscation of an SQL statement. It is only
+// collected when enabled via SQLConfig.
+type SQLMetadata struct {
+	// TablesCSV is a comma-separated list of tables that the query addresses.
+	TablesCSV string `json:"tables_csv"`
+	// Comments in a SQL statement.
```
Comment (gbbr) [inline pkg/obfuscate/obfuscate.go:138]:
```suggestion
	// Commands {holds,is} ...
```

File: pkg/obfuscate/obfuscate.go [hunk 9]
```diff
@@ -124,6 +130,17 @@ type SQLConfig struct {
...
+	Commands []string `json:"commands"`
+	// Comments in a SQL statement.
```
Comment (gbbr) [inline pkg/obfuscate/obfuscate.go:140]:
```suggestion
	// Comments holds ... when enabled via ...
```

File: pkg/obfuscate/obfuscate.go [hunk 10]
```diff
@@ -105,7 +105,13 @@ type StatsClient interface {
 type SQLConfig struct {
 	// TableNames specifies whether the obfuscator should also extract the table names that a query addresses,
 	// in addition to obfuscating.
-	TableNames bool
+	TableNames bool `json:"table_names"`
+
+	// CollectCommands specifies whether the obfuscator should extract and return commands as SQL metadata when obfuscating.
```
Comment (gbbr) [inline pkg/obfuscate/obfuscate.go:110]: Do we need to be more clear here on what a "command" is? Or is "command" a valid SQL spec term?

File: pkg/obfuscate/sql.go (hunk 7 above)
Comment (gbbr) [reply pkg/obfuscate/sql.go:210]: This change seems unrelated. Also, yes, this whole logic is quite hard to follow 😄

**Review by gbbr — COMMENTED** [review]
This is looking good 🙂 Please ping me whenever you like the final review & approval. I don't have any large comments, but I'd like to glance one last time before you feel you're ready.

File: pkg/obfuscate/sql.go (hunk 7 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go:210]: Sorry, for not getting around to replying sooner. Before I added the `Select` token [as a keyword](https://github.com/DataDog/datadog-agent/pull/10042/files#diff-c22948a50c7f268987e3e4bc260323ae02b0c5c639538cf2df173aab640d5db0), the SELECT command used to fall under the `ID` token which is essentially like an "other" category that can contain various other SQL values and commands that we have not yet specified. This check here is looking for subqueries to group, and subqueries begin with SELECT. Because I added the `Select` token, the tokenizer is now generating SELECT under the `Select` token, not `ID`, so this check here would never trigger.

File: pkg/obfuscate/obfuscate.go (hunk 10 above)
Comment (alexbarksdale) [reply pkg/obfuscate/obfuscate.go:110]: I believe SQL commands is a valid SQL spec term, but I can add a quick clarifying comment for more context.

File: pkg/obfuscate/obfuscate.go (hunk 9 above)
Comment (alexbarksdale) [reply pkg/obfuscate/obfuscate.go]: https://github.com/DataDog/datadog-agent/pull/10042/commits/f79b990dca3b4d04d4034f7442467ce711c14553

File: pkg/obfuscate/obfuscate.go (hunk 8 above)
Comment (alexbarksdale) [reply pkg/obfuscate/obfuscate.go]: https://github.com/DataDog/datadog-agent/pull/10042/commits/f79b990dca3b4d04d4034f7442467ce711c14553

File: pkg/obfuscate/obfuscate.go (hunk 10 above)
Comment (alexbarksdale) [reply pkg/obfuscate/obfuscate.go:110]: https://github.com/DataDog/datadog-agent/pull/10042/commits/f79b990dca3b4d04d4034f7442467ce711c14553

Comment (gbbr) [reply pkg/obfuscate/obfuscate.go:110]: If it is, that's fine. No need too. Thanks 🙏🏻

File: pkg/obfuscate/obfuscate.go [hunk 11]
```diff
@@ -124,6 +130,18 @@ type SQLConfig struct {
 	Cache bool
 }
 
+// SQLMetadata holds metadata collected throughout the obfuscation of an SQL statement. It is only
+// collected when enabled via SQLConfig.
+type SQLMetadata struct {
+	// TablesCSV is a comma-separated list of tables that the query addresses.
+	TablesCSV string `json:"tables_csv"`
+	// Commands holds commands executed in an SQL statement.
+	// e.g. SELECT, UPDATE, INSERT, DELETE, etc.
```
Comment (gbbr) [inline pkg/obfuscate/obfuscate.go:139]: Awesome! Thanks for adding this!

File: pkg/obfuscate/sql.go (hunk 7 above)
Comment (gbbr) [reply pkg/obfuscate/sql.go:217]: I see. It makes sense. But couldn't other identifiers/commands start a sub-query? Wouldn't we be playing it safer if we kept both? I mean `token == Select || token == ID` and even `token == Delete`.

**Review by gbbr — COMMENTED** [review]
This is really nice work Alex. Left one question.

File: pkg/obfuscate/sql.go (hunk 7 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go:217]: I did some research before hand regarding SQL grammar and couldn't find anything to suggest that subqueries can be with things like UPDATE or DELETE. However, you're right, we should prepare for the worst regardless and include the "catch all" aspect of this. Sending up a fix, ty @gbbr!

**Comment by alexbarksdale** [comment]
Hey @Taiki-San, tagging you in response to the easyjson discussion we had earlier this week. The changes in this PR require another regeneration. Please take a look when you get a chance to make sure we don't run into any conflicts. Happy to work with you to resolve any issues if any.

**Review by Taiki-San — APPROVED** [review]

File: pkg/obfuscate/sql.go [hunk 12]
```diff
@@ -16,9 +16,99 @@ import (
...
+}
+
+func (f *metadataFinderFilter) Filter(token, lastToken TokenKind, buffer []byte) (TokenKind, []byte, error) {
+	if f.collectComments && token == Comment {
+		// A comment with line-breaks will be brought to a single line.
+		f.comments = append(f.comments, strings.TrimSpace(strings.Replace(string(buffer), "\n", " ", -1)))
+	}
+	if f.collectCommands {
+		switch token {
+		case Select, Update, Insert, Delete, Join, Alter, Drop, Create, Grant, Revoke, Commit, Begin, Truncate:
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:44]: Awesome!

File: pkg/obfuscate/sql.go [hunk 13]
```diff
@@ -16,9 +16,99 @@ import (
...
+	return SQLMetadata{
+		TablesCSV: f.tablesCSV.String(),
+		Commands:  f.commands,
+		Comments:  f.comments,
+	}
+}
+
+// Reset implements tokenFilter.
+func (f *metadataFinderFilter) Reset() {
+	for k := range f.tablesSeen {
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:101]: I think we should reset everything here, right (including collected commands & comments) ? A bit concerning that the tests didn't catch this.

File: pkg/obfuscate/sql.go [hunk 14]
```diff
@@ -220,90 +313,35 @@ func (o *Obfuscator) obfuscateSQLString(in string, opts *SQLConfig) (*Obfuscated
...
-	TablesCSV string // comma-separated list of tables that the query addresses
+	Query    string      `json:"query"`    // the obfuscated SQL query
+	Metadata SQLMetadata `json:"metadata"` // metadata extracted from the SQL query
 }
 
 // Cost returns the number of bytes needed to store all the fields
 // of this ObfuscatedQuery.
 func (oq *ObfuscatedQuery) Cost() int64 {
-	return int64(len(oq.Query) + len(oq.TablesCSV))
+	return int64(len(oq.Query) + len(oq.Metadata.TablesCSV))
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:325]: This becomes invalid now. It should contain the length of all the strings now (including comments & commands). This method is called when performance is critical and a cache is used, so iterating here over two slices will not be a good idea. Instead, it would be nice to store the cost inside `SQLMetadata` and increase it dynamically when comments & commands are added. To do this nicely, I'd recommend removing the fields from `metadataFinderFilter` and having just one `SQLMetadata` field which stores all the information, having methods to add comments & commands, which in turn automatically increase the cost. Within that context, it will be better to call it `Size` instead of `Cost`.

**Review by gbbr — CHANGES_REQUESTED** [review]
Unfortunately we've missed some things the first time around...

File: pkg/obfuscate/sql.go (hunk 14 above)
Comment (gbbr) [reply pkg/obfuscate/sql.go:325]: How about computing the cost on the fly? I mean simply leaving things as they are, and adding a new field into the filter called `resultBytes` or something, and setting that as `SQLMetadata.Cost` at the end.

File: pkg/obfuscate/sql.go [hunk 15]
```diff
@@ -16,9 +16,114 @@ import (
...
 
+// metadataFinderFilter is a filter which attempts to collect metadata from a query, such as comments and tables.
+// It is meant to run before all the other filters.
+type metadataFinderFilter struct {
+	collectTableNames bool
+	collectCommands   bool
+	collectComments   bool
+	replaceDigits     bool
+
+	// size holds the byte size of the metadata collected by the filter.
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:27]: This is wrong (documentation wise) because it excludes the tablesCSV size. IMHO we should include it, just for the sake integrity.

File: pkg/obfuscate/sql.go [hunk 16]
```diff
@@ -16,9 +16,114 @@ import (
...
+		return
+	}
+	if f.tablesSeen == nil {
+		f.tablesSeen = make(map[string]struct{}, 1)
+	}
+	f.tablesSeen[name] = struct{}{}
+	if f.tablesCSV.Len() > 0 {
+		f.tablesCSV.WriteByte(',')
+	}
+	f.tablesCSV.WriteString(name)
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:93]: Should probably increase `f.size` here.

File: pkg/obfuscate/sql.go [hunk 17]
```diff
@@ -16,9 +16,114 @@ import (
...
+
+// Reset implements tokenFilter.
+func (f *metadataFinderFilter) Reset() {
+	for k := range f.tablesSeen {
+		delete(f.tablesSeen, k)
+	}
+	f.size = 0
+	f.tablesCSV.Reset()
+	f.commands = nil
+	f.comments = nil
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:119]:
```suggestion
	f.commands = f.commands[:0]
	f.comments = f.comments[:0]
```
This will keep the underlying array for the subsequent calls. Otherwise, with `nil`, we keep garbage collecting and re-allocating on every run, even though we could reuse what we started. WDYT?

**Review by gbbr — COMMENTED** [review]
This is great! Thanks. A few more things.

File: pkg/obfuscate/sql.go (hunk 15 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go:27]: https://github.com/DataDog/datadog-agent/pull/10042/commits/6be76eac6b841793ddd5162f904ee8b1557c4f2c

File: pkg/obfuscate/sql.go (hunk 16 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go:95]: https://github.com/DataDog/datadog-agent/pull/10042/commits/6be76eac6b841793ddd5162f904ee8b1557c4f2c

File: pkg/obfuscate/sql.go [hunk 18]
```diff
@@ -16,9 +16,111 @@ import (
...
+func (f *metadataFinderFilter) storeTableName(name string) {
+	if _, ok := f.tablesSeen[name]; ok {
+		return
+	}
+	if f.tablesSeen == nil {
+		f.tablesSeen = make(map[string]struct{}, 1)
+	}
+	f.tablesSeen[name] = struct{}{}
+	if f.tablesCSV.Len() > 0 {
+		f.size += 1
```
Comment (gbbr) [inline pkg/obfuscate/sql.go:91]: You might get a CI failure otherwise
```suggestion
		f.size++
```

**Review by gbbr — APPROVED** [review]
Thanks a lot Alex! Really nice work!

File: pkg/obfuscate/sql.go (hunk 18 above)
Comment (alexbarksdale) [reply pkg/obfuscate/sql.go]: hehe, just saw it fail on that, fixed!

**Review by ruthnaebeck — APPROVED** [review]
👍 for docs

File: pkg/obfuscate/obfuscate.go [hunk 19]
```diff
@@ -105,11 +105,20 @@ type StatsClient interface {
...
+	CollectCommands bool `json:"collect_commands"`
+
+	// CollectComments specifies whether the obfuscator should extract and return comments as SQL metadata when obfuscating.
+	CollectComments bool `json:"collect_comments"`
 
 	// ReplaceDigits specifies whether digits in table names and identifiers should be obfuscated.
 	ReplaceDigits bool `json:"replace_digits"`
 
+	// ReturnJSONMetadata specifies whether the python obfuscator stub will return metadata as JSON.
+	ReturnJSONMetadata bool `json:"return_json_metadata"`
```
Comment (gbbr) [inline pkg/obfuscate/obfuscate.go:120]: I'm afraid that I can't approve this change. It has nothing to do with the `pkg/obfuscate` module. It has no effect on obfuscation, and there is no concept of a "python obfuscator stub" in this package. Please find another way to accomplish this without a change to this module. Perhaps an additional function argument somewhere else.

**Review by gbbr — CHANGES_REQUESTED** [review]

File: pkg/obfuscate/obfuscate.go (hunk 19 above)
Comment (alexbarksdale) [reply pkg/obfuscate/obfuscate.go]: Updated: https://github.com/DataDog/datadog-agent/pull/10042/commits/b2101aa32ba4db16666edcad493c328063c21283

File: releasenotes/notes/add-option-to-collect-sql-metadata-9b22dd3f1fe30433.yaml [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    Add the option to collect SQL comments and commands during obfuscation.
```
Comment (gbbr) [inline releasenotes/notes/add-option-to-collect-sql-metadata-9b22dd3f1fe30433.yaml:11]:
```suggestion
    APM: Add the option to collect SQL comments and commands during obfuscation.
```
We need to add this prefix for trace-agent.

**Review by gbbr — APPROVED** [review]
Very nice work Alex! Thanks!

**Comment by KSerrania** [comment]
The Appveyor build in `pending` state succeeded, but its status wasn't reported back to GitHub. GIven that all other tests succeeded and all required approvals are here, I'll use my admin powers to merge the PR.
